### PR TITLE
Spark SQL driver: read Hive database schema and partitionned tables

### PR DESCRIPTION
**Pull request by jf-beauvais**
Related to bug: #7630

When reading from Hive server, the **show tables** query result don't have the same column name: **tab_name** (instead of tablename for Spark thrift server).

Additionally, the query result may contains empty rows or comments failing the describe-table parsing.

### CODE REVIEW

File: modules/drivers/sparksql/src/metabase/driver/sparksql.clj [hunk 1]
```diff
@@ -88,12 +88,19 @@
   (when s
     (s/replace s #"-" "_")))
 
+(defn- valid-describe-table-row [row]
```
Comment (camsaul) [inline modules/drivers/sparksql/src/metabase/driver/sparksql.clj]: https://github.com/bbatsov/clojure-style-guide#pred-with-question-mark

File: modules/drivers/sparksql/src/metabase/driver/sparksql.clj [hunk 2]
```diff
@@ -88,12 +88,19 @@
...
+       (s/blank? (:data_type row))
+       (s/starts-with? (:data_type row) "#"))))
+
 ;; workaround for SPARK-9686 Spark Thrift server doesn't return correct JDBC metadata
 (defmethod driver/describe-database :sparksql [_ {:keys [details] :as database}]
   {:tables (with-open [conn (jdbc/get-connection (sql-jdbc.conn/db->pooled-connection-spec database))]
              (set (for [result (jdbc/query {:connection conn}
                                            ["show tables"])]
-                    {:name   (:tablename result)
+                    {:name   (if (contains? result :tablename) (:tablename result) (:tab_name result)) ;; column name differ depending on server (SparkSQL, hive, Impala)
```
Comment (camsaul) [inline modules/drivers/sparksql/src/metabase/driver/sparksql.clj]: I would use (`some`)[https://clojuredocs.org/clojure.core/some] here

```clj
(some result [:tablename :tab_name])
```

to get either `:tablename` or `:tab_name` (whichever is truthy, if any)

File: modules/drivers/sparksql/src/metabase/driver/sparksql.clj [hunk 3]
```diff
@@ -88,12 +88,19 @@
   (when s
     (s/replace s #"-" "_")))
 
+(defn- valid-describe-table-row [row]
+  (not(or
```
Comment (camsaul) [inline modules/drivers/sparksql/src/metabase/driver/sparksql.clj]: missing space, should be `(not (or`

**Review by camsaul — APPROVED** [review]
~Thanks for the PR. If you fix the style issues we can go ahead and merge this~

EDIT: meant to mark this as "request changes"

**Review by camsaul — CHANGES_REQUESTED** [review]
Thanks for the PR. If you fix the style issues we can go ahead and merge this

**Comment by jf-beauvais** [comment]
> Thanks for the PR. If you fix the style issues we can go ahead and merge this

Thanks for your recommendations, it's much better this way.

File: modules/drivers/sparksql/src/metabase/driver/sparksql.clj (hunk 1 above)
Comment (camsaul) [reply modules/drivers/sparksql/src/metabase/driver/sparksql.clj]: If you really want to make this function spicy you can rewrite as follows 🌶 :

```clj
(defn- valid-describe-table-row? [{:keys [col_name data_type]}]
  (every? (every-pred (complement str/blank?)
                      (complement #(str/starts-with? % "#")))
          [col_name data_type]))
```

*  Change name to `valid-describe-table-row?` as I previously suggested in comment above
*  Destructure `row` to get `:col_name` and `:data_type` out in the arglist
*  Use `every?` which checks if a predicate function is true of every item in a collection
*  Use `every-pred` to construct a predicate function that is true if all predicate functions passed to it are true
*  Use `complement` to return a predicate function that returns the complement (i.e., opposite boolean value) of the original predicate function

Comment (jf-beauvais) [reply modules/drivers/sparksql/src/metabase/driver/sparksql.clj]: Sorry I missed it renaming the function with the question mark.
Thank you for the cool tips on Clojure.

**Comment by jf-beauvais** [comment]
I closed/reopen this PR to relaunch tests because I commited with the wrong gitlab user (which I deleted BTW)

**Comment by jensenity** [comment]
Hi, If my Spark Thrift Server has multiple schemas(database), what should I write on my Metabase Spark SQL connection's `Database`?

So far, I'm able to just set one schema(database) per connection. Is there a way to describe all schema and all the tables in those schema with just 1 Spark SQL connection on Metabase? https://discourse.metabase.com/t/my-spark-sql-database-doesnt-show-all-my-hive-schemas/13170/3

Thanks in advance.
@camsaul

**Pull request merged**